<a href="https://colab.research.google.com/github/yaramirezte/TelecomX_LATAM/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#📌 Extracción

In [1]:
import pandas as pd

In [2]:
url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

In [3]:
datos_TelecomX = pd.read_json(url)

In [4]:
datos_TelecomX.sample(10)

customerID Churn                                           customer  \
2426  3391-JSQEW    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
1210  1729-VLAZJ    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
3733  5143-EGQFK   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
3939  5380-XPJNZ    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
5229  7133-VBDCG   Yes  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
4927  6729-FZWSY    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2049  2885-HIJDH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
65    0107-YHINA   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
4809  6585-WCEWR    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
1709  2429-AYKKO    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
2426     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1210  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
3733  {'PhoneService': 'No', 'MultipleLines': 'No ph...   
3939     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5229     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4927     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
2049    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
65      {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
4809     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1709     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                               internet  \
2426  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1210  {'InternetService': 'DSL', 'OnlineSecurity': '...   
3733  {'InternetService': 'DSL', 'OnlineSecurity': '...   
3939  {'InternetService': 'No', 'OnlineSecurity': 'N...   
5229  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4927  {'InternetService': 'No', 'OnlineSecurity': 'N...   
2049  {'InternetService': 'DSL', 'OnlineSecurity': '...   
65    {'InternetService': 'Fiber optic', 'OnlineSecu...   
4809  {'InternetService': 'No', 'OnlineSecurity': 'N...   
1709  {'InternetService': 'No', 'OnlineSecurity': 'N...   

                                                account  
2426  {'Contract': 'Month-to-month', 'PaperlessBilli...  
1210  {'Contract': 'One year', 'PaperlessBilling': '...  
3733  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3939  {'Contract': 'One year', 'PaperlessBilling': '...  
5229  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4927  {'Contract': 'One year', 'PaperlessBilling': '...  
2049  {'Contract': 'Two year', 'PaperlessBilling': '...  
65    {'Contract': 'Month-to-month', 'PaperlessBilli...  
4809  {'Contract': 'Two year', 'PaperlessBilling': '...  
1709  {'Contract': 'Two year', 'PaperlessBilling': '...

#🔧 Transformación

In [5]:
datos_TelecomX['customer'].iloc[0]

{'gender': 'Female',
 'SeniorCitizen': 0,
 'Partner': 'Yes',
 'Dependents': 'Yes',
 'tenure': 9}

El dataset presenta una estructura JSON anidada, donde varias columnas contienen diccionarios con información detallada del cliente, servicios contratados y datos de facturación.

In [7]:
datos_TelecomX.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [8]:
datos_TelecomX.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [10]:
datos_TelecomX.isnull().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [11]:
datos_TelecomX.isna().sum()

customerID    0
Churn         0
customer      0
phone         0
internet      0
account       0
dtype: int64

In [13]:
datos_TelecomX.map(type).head()

customerID          Churn        customer           phone  \
0  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
1  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
2  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
3  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   
4  <class 'str'>  <class 'str'>  <class 'dict'>  <class 'dict'>   

         internet         account  
0  <class 'dict'>  <class 'dict'>  
1  <class 'dict'>  <class 'dict'>  
2  <class 'dict'>  <class 'dict'>  
3  <class 'dict'>  <class 'dict'>  
4  <class 'dict'>  <class 'dict'>

In [14]:
# pd.json_normalize() Recorre cada diccionario, entra dentro de los diccionarios anidados, saca cada clave interna, crea columnas nuevas usando notación con puntos
# datos_TelecomX.to_dict(orient="records") - Convierte el DataFrame en una lista de diccionarios, uno por fila.

datos_TelecomX_normalizados = pd.json_normalize(datos_TelecomX.to_dict(orient="records"))


In [15]:
datos_TelecomX_normalizados.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
2579  3594-IVHJZ    No          Female                       0   
3253  4538-WNTMJ    No          Female                       0   
3451  4785-NKHCX    No            Male                       1   
4349  5968-VXZLG    No            Male                       0   
6303  8670-ERCJH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
2579               No                  No                2                Yes   
3253              Yes                 Yes               46                Yes   
3451               No                  No                5                Yes   
4349               No                  No                2                Yes   
6303               No                  No               72                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
2579                  No                       No  ...   No internet service   
3253                 Yes                       No  ...   No internet service   
3451                  No                       No  ...   No internet service   
4349                  No                      DSL  ...                    No   
6303                 Yes              Fiber optic  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
2579       No internet service  No internet service  No internet service   
3253       No internet service  No internet service  No internet service   
3451       No internet service  No internet service  No internet service   
4349                        No                   No                   No   
6303                       Yes                  Yes                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
2579      No internet service   Month-to-month                      Yes   
3253      No internet service         One year                       No   
3451      No internet service   Month-to-month                       No   
4349                       No   Month-to-month                      Yes   
6303                      Yes         Two year                       No   

          account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
2579               Mailed check                   20.45                  42.45  
3253               Mailed check                   24.95                 1165.9  
3451  Bank transfer (automatic)                   20.05                  94.15  
4349  Bank transfer (automatic)                   46.60                   87.9  
6303  Bank transfer (automatic)                  113.40                 8164.1  

[5 rows x 21 columns]

In [16]:
datos_TelecomX_normalizados.duplicated().sum()

np.int64(0)

In [17]:
datos_TelecomX_normalizados['account.Contract'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [18]:
datos_TelecomX_normalizados['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [19]:
datos_TelecomX_normalizados['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [20]:
datos_TelecomX_normalizados.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total'],
      dtype='object')

In [21]:
datos_TelecomX_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [21]:
datos_TelecomX_normalizados.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

El conjunto de datos presenta información demográfica de los clientes características de los servicios contratados y datos contractuales y financieros. Tambien se consultó el diccionario de datos proporcionado. A partir de esta revisión, se identificó la variable Churn como objetivo, y se seleccionaron como variables relevantes aquellas relacionadas con el perfil del cliente, los servicios contratados y las condiciones del contrato.

In [22]:
df = datos_TelecomX_normalizados.copy()

In [23]:
for col in df.select_dtypes(include="object").columns:
  df[col] = df[col].str.strip()

In [24]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [25]:
(df['Churn'] == '').sum()

np.int64(224)

In [26]:
churn_vacios = df[df['Churn'] == '']

In [27]:
churn_vacios

customerID Churn customer.gender  customer.SeniorCitizen  \
30    0047-ZHDTW                Female                       0   
75    0120-YZLQA                  Male                       0   
96    0154-QYHJU                  Male                       0   
98    0162-RZGMZ                Female                       1   
175   0274-VVQOQ                  Male                       1   
...          ...   ...             ...                     ...   
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
30                 No                  No               11                Yes   
75                 No                  No               71                Yes   
96                 No                  No               29                Yes   
98                 No                  No                5                Yes   
175               Yes                  No               65                Yes   
...               ...                 ...              ...                ...   
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
30                   Yes              Fiber optic  ...                    No   
75                    No                       No  ...   No internet service   
96                    No                      DSL  ...                   Yes   
98                    No                      DSL  ...                   Yes   
175                  Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
30                          No                   No                   No   
75         No internet service  No internet service  No internet service   
96                          No                  Yes                   No   
98                          No                  Yes                   No   
175                        Yes                   No                  Yes   
...                        ...                  ...                  ...   
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
30                         No   Month-to-month                      Yes   
75        No internet service         Two year                      Yes   
96                         No         One year                      Yes   
98                         No   Month-to-month                    

In [28]:
churn_vacios[['customer.tenure', 'account.Charges.Total', 'account.Charges.Monthly']].head()

customer.tenure account.Charges.Total  account.Charges.Monthly
30                11                 929.3                    79.00
75                71                1355.1                    19.90
96                29                1696.2                    58.75
98                 5                287.85                    59.90
175               65               6792.45                   103.15

In [29]:
# eliminar los registros con df['Churn'] vacios
df = df[df['Churn'] != '']

In [30]:
# df['Churn'] = df['Churn'].str.capitalize()

In [30]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

Se identificaron 224 registros con la variable objetivo Churn vacía. Al explorar estos clientes se observó que poseen historial de facturación y antigüedad variable (customer.tenure), pero no es posible inferir su estado de evasión. Por este motivo, estos registros fueron eliminados del conjunto de datos, asegurando la integridad de la variable objetivo.

In [31]:
df['account.Charges.Total'].dtype

dtype('O')

In [35]:
mask_no_numericos = ~df['account.Charges.Total'].astype(str).str.match(r'^\d+(\.\d+)?$')
df.loc[mask_no_numericos, 'account.Charges.Total']


975     
1775    
1955    
2075    
2232    
2308    
2930    
3134    
3203    
4169    
5599    
Name: account.Charges.Total, dtype: object

In [37]:
df.loc[mask_no_numericos, 'account.Charges.Total'].apply(
    lambda x: repr(x)
)

975     ''
1775    ''
1955    ''
2075    ''
2232    ''
2308    ''
2930    ''
3134    ''
3203    ''
4169    ''
5599    ''
Name: account.Charges.Total, dtype: object

In [36]:
mask_no_numericos.sum()

np.int64(11)

In [32]:
mask_no_numericos = pd.to_numeric(df['account.Charges.Total'], errors='coerce').isna() & df['account.Charges.Total'].notna()

df.loc[mask_no_numericos, 'account.Charges.Total']

975     
1775    
1955    
2075    
2232    
2308    
2930    
3134    
3203    
4169    
5599    
Name: account.Charges.Total, dtype: object

In [33]:
df.loc[mask_no_numericos, 'account.Charges.Total'].unique()

array([''], dtype=object)

In [34]:
df.loc[mask_no_numericos]


customerID Churn customer.gender  customer.SeniorCitizen  \
975   1371-DWPAZ    No          Female                       0   
1775  2520-SGTTA    No          Female                       0   
1955  2775-SEFEE    No            Male                       0   
2075  2923-ARZLG    No            Male                       0   
2232  3115-CZMZD    No            Male                       0   
2308  3213-VVOLG    No            Male                       0   
2930  4075-WKNIU    No          Female                       0   
3134  4367-NUYAO    No            Male                       0   
3203  4472-LVYGI    No          Female                       0   
4169  5709-LVOEQ    No          Female                       0   
5599  7644-OMVMY    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
975               Yes                 Yes                0                 No   
1775              Yes                 Yes                0                Yes   
1955               No                 Yes                0                Yes   
2075              Yes                 Yes                0                Yes   
2232               No                 Yes                0                Yes   
2308              Yes                 Yes                0                Yes   
2930              Yes                 Yes                0                Yes   
3134              Yes                 Yes                0                Yes   
3203              Yes                 Yes                0                 No   
4169              Yes                 Yes                0                Yes   
5599              Yes                 Yes                0                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
975     No phone service                      DSL  ...                   Yes   
1775                  No                       No  ...   No internet service   
1955                 Yes                      DSL  ...                   Yes   
2075                  No                       No  ...   No internet service   
2232                  No                       No  ...   No internet service   
2308                 Yes                       No  ...   No internet service   
2930                 Yes                      DSL  ...                   Yes   
3134                 Yes                       No  ...   No internet service   
3203    No phone service                      DSL  ...                    No   
4169                  No                      DSL  ...                   Yes   
5599                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
975                        Yes                  Yes                  Yes   
1775       No internet service  No internet service  No internet service   
1955                        No                  Yes                   No   
2075       No internet service  No internet service  No internet service   
2232       No internet service  No internet service  No internet service   
2308       No internet service  No internet service  No internet service   
2930                       Yes                  Yes                  Yes   
3134       No internet service  No internet service  No internet service   
3203                       Yes                  Yes                  Yes   
4169                       Yes                   No                  Yes   
5599       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
975                        No         Two year                       No   
1775      No internet service         Two year                       No   
1955                       No         Two year                      Yes   
2075      No internet service         One year                    

In [38]:
df['account.Charges.Total'] = df['account.Charges.Total'].replace(['', ' '],pd.NA) # Reemplazar strings vacíos por NaN


In [39]:
df['account.Charges.Total']= pd.to_numeric(df['account.Charges.Total'], errors='coerce') #Convertir a numérico

In [40]:
df['account.Charges.Total'].dtype

dtype('float64')

In [41]:
df.isnull().sum()

customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64

In [36]:
df.isna().sum()

customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64

In [42]:
df.loc[df['account.Charges.Total'].isnull(), ['account.Charges.Total', 'customer.tenure', 'account.Charges.Monthly']]

account.Charges.Total  customer.tenure  account.Charges.Monthly
975                     NaN                0                    56.05
1775                    NaN                0                    20.00
1955                    NaN                0                    61.90
2075                    NaN                0                    19.70
2232                    NaN                0                    20.25
2308                    NaN                0                    25.35
2930                    NaN                0                    73.35
3134                    NaN                0                    25.75
3203                    NaN                0                    52.55
4169                    NaN                0                    80.85
5599                    NaN                0                    19.85

Si un cliente acaba de entrar (tenure = 0), entonces su total de cargos debe ser 0, porque todavía no ha pagado nada.

In [44]:
df.loc[df['customer.tenure'] == 0, 'account.Charges.Total'] = 0 # Selecciona solo la columna account.Charges.Total solo en las filas donde la condición es verdadera y pone 0
# Esto corrige los valores vacíos o NaN para clientes nuevos.

In [45]:
df['account.Charges.Total'].isnull().sum()

np.int64(0)

In [46]:
df.isna().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

Se identificaron 11 valores nulos en la variable account.Charges.Total. Al analizar estos registros se comprobó que corresponden a clientes con antigüedad (customer.tenure) igual a cero, quienes aún no han generado cargos acumulados. Por este motivo, los valores nulos fueron actualizados con el valor 0, garantizando la coherencia de la información financiera.

In [42]:
df['internet.InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [43]:
df['internet.OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [44]:
# Consultar
df.loc[df['internet.InternetService'] == 'No',['internet.InternetService', 'internet.OnlineSecurity', 'internet.OnlineBackup']]


internet.InternetService internet.OnlineSecurity internet.OnlineBackup
20                         No     No internet service   No internet service
23                         No     No internet service   No internet service
24                         No     No internet service   No internet service
27                         No     No internet service   No internet service
28                         No     No internet service   No internet service
...                       ...                     ...                   ...
7250                       No     No internet service   No internet service
7252                       No     No internet service   No internet service
7256                       No     No internet service   No internet service
7257                       No     No internet service   No internet service
7261                       No     No internet service   No internet service

[1526 rows x 3 columns]

Se verificaron las variables relacionadas con servicios de internet y se observó que, para los clientes sin servicio de internet (InternetService = "No"), las variables de servicios adicionales toman el valor "No internet service". Esta codificación es consistente y representa correctamente la ausencia de dichos servicios, por lo que no se realizaron modificaciones en estas columnas.

In [47]:
df.duplicated().sum()

np.int64(0)

In [48]:
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7043 non-null   object 
 1   Churn                      7043 non-null   object 
 2   customer.gender            7043 non-null   object 
 3   customer.SeniorCitizen     7043 non-null   int64  
 4   customer.Partner           7043 non-null   object 
 5   customer.Dependents        7043 non-null   object 
 6   customer.tenure            7043 non-null   int64  
 7   phone.PhoneService         7043 non-null   object 
 8   phone.MultipleLines        7043 non-null   object 
 9   internet.InternetService   7043 non-null   object 
 10  internet.OnlineSecurity    7043 non-null   object 
 11  internet.OnlineBackup      7043 non-null   object 
 12  internet.DeviceProtection  7043 non-null   object 
 13  internet.TechSupport       7043 non-null   object 
 1

customerID Churn customer.gender  customer.SeniorCitizen  \
count         7043  7043            7043             7043.000000   
unique        7043     2               2                     NaN   
top     9995-HOTOH    No            Male                     NaN   
freq             1  5174            3555                     NaN   
mean           NaN   NaN             NaN                0.162147   
std            NaN   NaN             NaN                0.368612   
min            NaN   NaN             NaN                0.000000   
25%            NaN   NaN             NaN                0.000000   
50%            NaN   NaN             NaN                0.000000   
75%            NaN   NaN             NaN                0.000000   
max            NaN   NaN             NaN                1.000000   

       customer.Partner customer.Dependents  customer.tenure  \
count              7043                7043      7043.000000   
unique                2                   2              NaN   
top                  No                  No              NaN   
freq               3641                4933              NaN   
mean                NaN                 NaN        32.371149   
std                 NaN                 NaN        24.559481   
min                 NaN                 NaN         0.000000   
25%                 NaN                 NaN         9.000000   
50%                 NaN                 NaN        29.000000   
75%                 NaN                 NaN        55.000000   
max                 NaN                 NaN        72.000000   

       phone.PhoneService phone.MultipleLines internet.InternetService  ...  \
count                7043                7043                     7043  ...   
unique                  2                   3                        3  ...   
top                   Yes                  No              Fiber optic  ...   
freq                 6361                3390                     3096  ...   
mean                  NaN                 NaN                      NaN  ...   
std                   NaN                 NaN                      NaN  ...   
min                   NaN                 NaN                      NaN  ...   
25%                   NaN                 NaN                      NaN  ...   
50%                   NaN                 NaN                      NaN  ...   
75%                   NaN                 NaN                      NaN  ...   
max                   NaN                 NaN                      NaN  ...   

       internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
count                   7043                      7043                 7043   
unique                     3                         3                    3   
top                       No                        No                   No   
freq                    3088                      3095                 3473   
mean                     NaN                       NaN                  NaN   
std                      NaN                       NaN                  NaN   
min                      NaN                       NaN                  NaN   
25%                      NaN                       NaN                  NaN   
50%                      NaN                       NaN                  NaN   
75%                      NaN                       NaN                  NaN   
max                      NaN                       NaN                  NaN   

       internet.StreamingTV internet.StreamingMovies account.Contract  \
count                  7043                     7043             7043   
unique                    3                        3                3   
top                      No                       No   Month-to-month   
freq                   2810                     2785             3875   
mean                    NaN                      NaN              NaN   
std                     NaN                      NaN              NaN   
min          

In [49]:
# Crear la columna Cuentas_Diarias
df.loc[:,'account.Charges.Daily'] = (df['account.Charges.Monthly'] / 30).round(2)

In [50]:
df[['account.Charges.Daily', 'account.Charges.Monthly']].head(5)

account.Charges.Daily  account.Charges.Monthly
0                   2.19                     65.6
1                   2.00                     59.9
2                   2.46                     73.9
3                   3.27                     98.0
4                   2.80                     83.9

Se creó la columna Cuentas_Diarias dividiendo la facturación mensual entre 30 días, proporcionando una métrica de gasto diario por cliente. Esta nueva variable permite analizar de manera más granular el comportamiento de los clientes y detectar patrones de evasión en función del gasto diario.

In [52]:
for col in df.select_dtypes(include='object').columns:
    print(col, "→", df[col].unique(), "\n")

customerID → ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH'] 

Churn → ['No' 'Yes'] 

customer.gender → ['Female' 'Male'] 

customer.Partner → ['Yes' 'No'] 

customer.Dependents → ['Yes' 'No'] 

phone.PhoneService → ['Yes' 'No'] 

phone.MultipleLines → ['No' 'Yes' 'No phone service'] 

internet.InternetService → ['DSL' 'Fiber optic' 'No'] 

internet.OnlineSecurity → ['No' 'Yes' 'No internet service'] 

internet.OnlineBackup → ['Yes' 'No' 'No internet service'] 

internet.DeviceProtection → ['No' 'Yes' 'No internet service'] 

internet.TechSupport → ['Yes' 'No' 'No internet service'] 

internet.StreamingTV → ['Yes' 'No' 'No internet service'] 

internet.StreamingMovies → ['No' 'Yes' 'No internet service'] 

account.Contract → ['One year' 'Month-to-month' 'Two year'] 

account.PaperlessBilling → ['Yes' 'No'] 

account.PaymentMethod → ['Mailed check' 'Electronic check' 'Credit card (automatic)'
 'Bank transfer (automatic)'] 



In [53]:
# Convertir valores textuales en binarios
binarias = ['customer.Partner', 'customer.Dependents', 'phone.PhoneService',
            'phone.MultipleLines', 'internet.OnlineSecurity', 'internet.OnlineBackup',
            'internet.DeviceProtection', 'internet.TechSupport', 'internet.StreamingTV',
            'internet.StreamingMovies', 'account.PaperlessBilling']

In [54]:
# se transforma
for col in binarias:
    df.loc[:,col] = df[col].map({'Yes': 1, 'No': 0,'No internet service': 0, 'No phone service': 0 })

In [55]:
df.sample(5)

customerID Churn customer.gender  customer.SeniorCitizen  \
1738  2472-OVKUP   Yes            Male                       0   
1157  1641-RQDAY    No          Female                       1   
3058  4256-ZWTZI    No          Female                       0   
63    0106-UGRDO    No          Female                       0   
5514  7560-QJAVJ    No          Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
1738                1                   0                6                  1   
1157                1                   1               32                  1   
3058                0                   1                5                  1   
63                  1                   0               69                  1   
5514                0                   0               70                  1   

     phone.MultipleLines internet.InternetService  ...  \
1738                   0              Fiber optic  ...   
1157                   1              Fiber optic  ...   
3058                   0                      DSL  ...   
63                     1              Fiber optic  ...   
5514                   0                      DSL  ...   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
1738                         1                    0                    0   
1157                         0                    0                    1   
3058                         0                    1                    0   
63                           1                    1                    1   
5514                         1                    1                    1   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
1738                        1   Month-to-month                        1   
1157                        1   Month-to-month                        1   
3058                        0   Month-to-month                        1   
63                          1         Two year                        1   
5514                        1         Two year                        0   

          account.PaymentMethod account.Charges.Monthly account.Charges.Total  \
1738           Electronic check                    91.0                531.00   
1157           Electronic check                   100.4               3217.65   
3058               Mailed check                    51.0                286.80   
63             Electronic check                   116.0               8182.85   
5514  Bank transfer (automatic)                    79.6               5589.45   

      account.Charges.Daily  
1738                   3.03  
1157                   3.35  
3058                   1.70  
63                     3.87  
5514                   2.65  

[5 rows x 22 columns]

In [56]:
df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})

In [57]:
import numpy as np
# Verificamos mínimos y máximos
min_val = df['account.Charges.Daily'].min()
max_val = df['account.Charges.Daily'].max()

In [58]:
# Creamos 5 bins equidistantes entre min y max
bins = np.linspace(min_val, max_val, 6)  # 6 límites = 5 segmentos
labels = ['Very Low', 'Low', 'Medium', 'High', 'Very High']

In [59]:
# Crear columna segmentada
df['Daily_Segment'] = pd.cut(
    df['account.Charges.Daily'],
    bins=bins,
    labels=labels,
    include_lowest=True  # Incluye el valor mínimo en el primer bin
)


In [60]:
# Verificamos el resultado
print(df[['account.Charges.Daily', 'Daily_Segment']].head(10))
print("\nDistribución de segmentos:")
print(df['Daily_Segment'].value_counts())

   account.Charges.Daily Daily_Segment
0                   2.19        Medium
1                   2.00        Medium
2                   2.46        Medium
3                   3.27          High
4                   2.80          High
5                   2.31        Medium
6                   3.66     Very High
7                   2.82          High
8                   1.61           Low
9                   3.02          High

Distribución de segmentos:
Daily_Segment
High         1827
Very Low     1801
Medium       1371
Very High    1033
Low          1011
Name: count, dtype: int64


In [61]:
df.columns

Index(['customerID', 'Churn', 'customer.gender', 'customer.SeniorCitizen',
       'customer.Partner', 'customer.Dependents', 'customer.tenure',
       'phone.PhoneService', 'phone.MultipleLines', 'internet.InternetService',
       'internet.OnlineSecurity', 'internet.OnlineBackup',
       'internet.DeviceProtection', 'internet.TechSupport',
       'internet.StreamingTV', 'internet.StreamingMovies', 'account.Contract',
       'account.PaperlessBilling', 'account.PaymentMethod',
       'account.Charges.Monthly', 'account.Charges.Total',
       'account.Charges.Daily', 'Daily_Segment'],
      dtype='object')

In [62]:
df = df.rename(columns={
    'customerID': 'ID_Cliente',
    'Churn': 'Evasion',
    'customer.gender': 'Genero',
    'customer.SeniorCitizen': 'Adulto_Mayor',
    'customer.Partner': 'Tiene_Pareja',
    'customer.Dependents': 'Tiene_Dependientes',
    'customer.tenure': 'Antiguedad_Meses',
    'phone.PhoneService': 'Servicio_Telefonico',
    'phone.MultipleLines': 'Multiples_Lineas',
    'internet.InternetService': 'Tipo_Internet',
    'internet.OnlineSecurity': 'Seguridad_Online',
    'internet.OnlineBackup': 'Backup_Online',
    'internet.DeviceProtection': 'Proteccion_Dispositivo',
    'internet.TechSupport': 'Soporte_Tecnico',
    'internet.StreamingTV': 'Streaming_TV',
    'internet.StreamingMovies': 'Streaming_Peliculas',
    'account.Contract': 'Tipo_Contrato',
    'account.PaperlessBilling': 'Factura_Digital',
    'account.PaymentMethod': 'Metodo_Pago',
    'account.Charges.Monthly': 'Cargos_Mensuales',
    'account.Charges.Total': 'Cargos_Totales',
    'account.Charges.Daily': 'Cargos_Diarios',
    'Daily_Segment': 'Segmento_Gasto_Diario'
})


In [61]:
# analizar churn por segmento
#df.groupby('Daily_Segment', observed=True)['Churn'].mean()

#📊 Carga y análisis

In [63]:
df.describe()

Evasion  Adulto_Mayor  Antiguedad_Meses  Cargos_Mensuales  \
count  7043.000000   7043.000000       7043.000000       7043.000000   
mean      0.265370      0.162147         32.371149         64.761692   
std       0.441561      0.368612         24.559481         30.090047   
min       0.000000      0.000000          0.000000         18.250000   
25%       0.000000      0.000000          9.000000         35.500000   
50%       0.000000      0.000000         29.000000         70.350000   
75%       1.000000      0.000000         55.000000         89.850000   
max       1.000000      1.000000         72.000000        118.750000   

       Cargos_Totales  Cargos_Diarios  
count     7043.000000     7043.000000  
mean      2279.734304        2.158675  
std       2266.794470        1.003088  
min          0.000000        0.610000  
25%        398.550000        1.180000  
50%       1394.550000        2.340000  
75%       3786.600000        2.990000  
max       8684.800000        3.960000

In [64]:
df['Evasion'].value_counts(normalize=True)

Evasion
0    0.73463
1    0.26537
Name: proportion, dtype: float64

In [66]:
# Contar clientes por estado de Evasion
conteo_evasion = df['Evasion'].value_counts().reset_index()
conteo_evasion

Evasion  count
0        0   5174
1        1   1869

In [67]:
conteo_evasion.columns = ['Evasion', 'Cantidad']
conteo_evasion

Evasion  Cantidad
0        0      5174
1        1      1869

In [68]:
conteo_evasion['Estado'] = conteo_evasion ['Evasion'].map({
    0: 'Clientes que permanecen',
    1: 'Clientes que se dieron de baja'
})

In [69]:
conteo_evasion

Evasion  Cantidad                          Estado
0        0      5174         Clientes que permanecen
1        1      1869  Clientes que se dieron de baja

In [70]:
# Gráfico tipo dona para representar el conteo de evasion
import plotly.express as px
fig = px.pie(
    conteo_evasion,
    values='Cantidad',
    names='Estado',
    hole=0.5,
    title='Distribución de clientes según evasión',
    color='Estado',
    color_discrete_map={
        'Clientes que permanecen': '#2ca02c',
        'Clientes que se dieron de baja': '#d62728'
    }
)


fig.update_traces(
    textinfo='percent+label',
    textfont_size=14,
    pull=[0, 0.05]  # Resalta ligeramente la Evasion
)

fig.update_layout(
    width=600,
    height=450,
    font_family='Arial',
    font_size=14,
    legend_title_text='',
    showlegend=True
)

fig.show()


El gráfico muestra la proporción de clientes que permanecen frente a aquellos que se dieron de baja. Se observa que la mayoría de los clientes continúa con el servicio, aunque existe un porcentaje relevante de evasión que representa una oportunidad de mejora para estrategias de retención.

In [70]:
df['Evasion'].mean()

np.float64(0.2653698707936959)

In [72]:
df.groupby('Genero')['Evasion'].mean()

Genero
Female    0.269209
Male      0.261603
Name: Evasion, dtype: float64

In [68]:
df['Tipo_Contrato'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [73]:
df.groupby('Tipo_Contrato')['Evasion'].mean().sort_values(ascending=False)

Tipo_Contrato
Month-to-month    0.427097
One year          0.112695
Two year          0.028319
Name: Evasion, dtype: float64

Contratos mensuales tienen mucha mayor evasión

In [74]:
df.groupby('Metodo_Pago')['Evasion'].mean().sort_values(ascending=False)

Metodo_Pago
Electronic check             0.452854
Mailed check                 0.191067
Bank transfer (automatic)    0.167098
Credit card (automatic)      0.152431
Name: Evasion, dtype: float64

Cheque electrónico tiene mayor evasión y Débito automático tiene la menor evasión

In [75]:
# creación de funcion para generar graficos de Evasion por genero, según tipo de contrato y metodo de pago
def grafico_evasion_categoria(df, columna, titulo):
  tasa = (df.groupby(columna)['Evasion'].mean().reset_index())

  max_valor = tasa['Evasion'].max()

  fig = px.bar(tasa, x= columna, y= 'Evasion', text = (tasa['Evasion']*100).round(1).astype(str) + '%', title=titulo
               )

  fig.update_layout(
      yaxis = dict(tickformat='.0%', range=[0, max_valor * 1.15]), # espacio extra
      yaxis_title='Tasa de Evasion (%)',
      xaxis_title=columna,
      width=780,
      height=450,
      font_family='Arial',
      font_size=14,
      plot_bgcolor='rgba(245,248,250,1)'
      #xaxis_tickangle=-20
  )
  fig.update_traces(textposition = 'outside', cliponaxis=False)
  fig.show()

In [80]:
grafico_evasion_categoria(
    df,
    'Genero',
    'Tasa de evasión según género'
)

In [77]:
grafico_evasion_categoria(
    df,
    'Tipo_Contrato',
    'Tasa de evasión según tipo de contrato'
)

In [78]:
grafico_evasion_categoria(
    df,
    'Metodo_Pago',
    'Tasa de evasión según método de pago'
)

In [ ]:
'''
def grafico_evasion_categoria_colores(df, columna, titulo):
    tasa = df.groupby(columna)['Evasion'].mean().reset_index()

    max_valor = tasa['Evasion'].max()

    fig = px.bar(
        tasa,
        x=columna,
        y='Evasion',
        text=(tasa['Evasion']*100).round(1).astype(str)+'%',
        title=titulo,
        color='Evasion',                  # gradiente según valor
        color_continuous_scale='RdYlGn_r' # rojo → verde
    )

    fig.update_layout(
        yaxis_title='Tasa de evasión (%)',
        xaxis_title=columna,
        yaxis=dict(
            tickformat='.0%',
            range=[0, max_valor*1.15]
        ),
        width=780,
        height=450,
        font_family='Arial',
        font_size=14,
        plot_bgcolor='rgba(245,248,250,1)',
        showlegend=False
    )

    fig.update_traces(
        textposition='outside',
        cliponaxis=False
    )

    fig.show()  # para usar despues quizas '''


In [97]:
# grafico de caja para demostrar la Antigüedad vs evasión
fig = px.box(df, x ='Evasion', y = 'Antiguedad_Meses', color='Evasion', title='Antigüedad del cliente según evasión',
             labels={'Evasion': 'Evasion (0 = Permanece, 1 = Cancela)', 'Antiguedad_Meses': 'Antigüedad (meses)'})
fig.update_layout(
    width=760,
    height=450,
    font_family='Arial',
    font_size=14,
    plot_bgcolor='rgba(245,248,250,1)',
    showlegend=False
)
fig.show()

Clientes con menor antigüedad presentan mayor evasión.

In [98]:
#Cargos totales vs evasión

fig = px.box(df, x ='Evasion', y = 'Cargos_Totales', color='Evasion', title='Total gastado según evasión',
             labels = {'Evasion': 'Evasion (0 = Permanece, 1 = Cancela)', 'Cargos_Totales':'Total gastado'})
fig.update_layout(
    width=760,
    height=450,
    font_family='Arial',
    font_size=14,
    plot_bgcolor='rgba(245,248,250,1)',
    showlegend=False
)
fig.show()

Clientes que cancelan han acumulado menor gasto total.

In [99]:
# Cargos mensuales
fig = px.box(df, x ='Evasion', y = 'Cargos_Mensuales', color='Evasion', title='Cargos mensuales según evasión',
             labels = {'Evasion': 'Evasion (0 = Permanece, 1 = Cancela)', 'Cargos_Mensuales':'Cargos mensuales'})
fig.update_layout(
    width=760,
    height=450,
    font_family='Arial',
    font_size=14,
    plot_bgcolor='rgba(245,248,250,1)',
    showlegend=False
)
fig.show()

Clientes que cancelan han tenido mayor gasto mensual.

El análisis de las variables numéricas muestra que los clientes que cancelaron el servicio presentan, en promedio, una menor antigüedad y un menor gasto total acumulado. Esto sugiere que la evasión ocurre principalmente en las primeras etapas de la relación con el cliente, lo que representa una oportunidad para implementar estrategias de retención temprana.

#📄Informe final

## Introducción

La evasión de clientes (churn) es uno de los principales desafíos para las empresas de telecomunicaciones, ya que la pérdida de clientes impacta directamente en los ingresos y en la sostenibilidad del negocio.
El objetivo de este análisis es comprender los factores asociados a la evasión de clientes en Telecom X, utilizando datos demográficos, características del servicio y variables de facturación.

A través de un análisis exploratorio de datos, se busca identificar patrones y perfiles de clientes con mayor probabilidad de evasión, con el fin de apoyar la toma de decisiones y proponer estrategias orientadas a la retención.

## Limpieza y Tratamiento de Datos

Los datos fueron obtenidos directamente desde una API en formato JSON y procesados utilizando Python y la librería pandas. Durante esta etapa se realizaron las siguientes acciones:

- Carga de los datos desde la API y normalización de la estructura JSON.

- Exploración de la estructura del dataset y verificación de tipos de datos.

- Identificación y tratamiento de valores nulos, especialmente en la variable de cargos totales.

- Corrección de inconsistencias en variables categóricas (por ejemplo, servicios no disponibles).

- Eliminación de registros con información incompleta en la variable objetivo (evasión).

- Creación de nuevas variables, como Cargos_Diarios, para enriquecer el análisis.

- Renombrado y traducción de columnas para mejorar la claridad y comprensión del dataset.

- Transformación de variables categóricas binarias a valores numéricos (0 y 1).


Como resultado, se obtuvo un conjunto de datos limpio, coherente y listo para el análisis exploratorio.

## Análisis Exploratorio de Datos
### Distribución de la evasión

Se analizó la proporción de clientes que permanecen frente a aquellos que cancelaron el servicio. La visualización permitió identificar que, aunque la mayoría de los clientes permanece, existe una proporción significativa de evasión que justifica un análisis más profundo.

### Evasión según variables categóricas

El análisis por género, tipo de contrato y método de pago reveló patrones importantes:

- Los clientes con contratos mes a mes presentan una mayor tasa de evasión en comparación con contratos de mayor duración.

- Determinados métodos de pago, como el pago electrónico, están asociados a una mayor evasión.

- No se observaron diferencias muy marcadas por género, lo que sugiere que este factor no es determinante por sí solo.

### Evasión y variables numéricas

Al analizar variables numéricas como la antigüedad y los cargos:

- Los clientes que cancelan el servicio suelen tener menor antigüedad, lo que indica que la evasión ocurre principalmente en las primeras etapas de la relación con la empresa.

- Los clientes que evaden también presentan, en promedio, un menor gasto total acumulado.

- La segmentación por gasto diario mostró que ciertos niveles de consumo están asociados a mayores tasas de evasión.

Estas visualizaciones permitieron identificar relaciones claras entre el comportamiento de los clientes y la probabilidad de evasión.

## Conclusiones e Insights

A partir del análisis realizado, se destacan los siguientes hallazgos clave:

- La evasión está fuertemente asociada a clientes con baja antigüedad.

- Los contratos de corto plazo presentan una mayor propensión al churn(Evasion).

- El comportamiento de pago y el nivel de gasto influyen en la probabilidad de cancelación.

- La evasión no depende de un solo factor, sino de una combinación de características del cliente y del servicio.

- Estos insights permiten comprender mejor el perfil de los clientes con mayor riesgo de evasión.

## Recomendaciones

Con base en los resultados obtenidos, se proponen las siguientes acciones estratégicas:

- Implementar programas de retención temprana dirigidos a clientes nuevos, especialmente durante los primeros meses de contrato.

- Incentivar contratos de mayor duración mediante descuentos o beneficios adicionales.

- Revisar y optimizar los métodos de pago asociados a mayores tasas de evasión.

- Diseñar ofertas personalizadas para segmentos con alto riesgo de churn(Evasion), basadas en su nivel de gasto y uso de servicios.

- Monitorear de forma continua la evasión para detectar cambios en el comportamiento de los clientes.

## Cierre

Este análisis proporciona una base sólida para la toma de decisiones orientadas a reducir la evasión de clientes. A partir de estos resultados, la empresa puede desarrollar estrategias más efectivas de fidelización y mejorar la experiencia del cliente, impactando positivamente en la rentabilidad y sostenibilidad del negocio.

# ¡Extra! Análisis de correlación entre variables

In [77]:
# Seleccionamos solo las variables numéricas
numericas = ['Antiguedad_Meses', 'Cargos_Mensuales', 'Cargos_Totales', 'Cargos_Diarios', 'Evasion']

# Calculamos la matriz de correlación
corr_matrix = df[numericas].corr()

# Mostramos la correlación con Evasion
print("Correlación con Evasion:\n", corr_matrix['Evasion'].sort_values(ascending=False))


Correlación con Evasion:
 Evasion             1.000000
Cargos_Diarios      0.193412
Cargos_Mensuales    0.193356
Cargos_Totales     -0.198324
Antiguedad_Meses   -0.352229
Name: Evasion, dtype: float64


In [79]:
# ver cómo los cargos diarios se relacionan con la evasión
fig = px.box(
    df,
    x='Evasion',
    y='Cargos_Diarios',
    color='Evasion',
    title='Relación entre Cargos Diarios y Evasión',
    labels={'Evasion':'Evasión (0 = Permanece, 1 = Cancela)', 'Cargos_Diarios':'Cargos diarios'}
)
fig.show()

**Insight:**

 Mayor gasto diario se asocia con mayor evasión. Clientes con cargos más altos tienen más probabilidad de abandonar.

Podría ser un indicador de que el costo percibido es un factor de abandono.

In [80]:
# cantidad de servicios contratados vs evasión

# Contar servicios contratados por cada cliente (solo 1 si tienen el servicio)
servicios = ['Servicio_Telefonico', 'Multiples_Lineas', 'Seguridad_Online',
             'Backup_Online', 'Proteccion_Dispositivo', 'Soporte_Tecnico',
             'Streaming_TV', 'Streaming_Peliculas']

In [83]:
df[servicios].sum(axis=1)
#axis=1 suma horizontalmente cada fila, si encuentra 1 suma y si encuentra 0 no suma y asi pasa por esas columnas

0       4
1       3
2       2
3       5
4       3
       ..
7262    3
7263    3
7264    2
7265    5
7266    5
Length: 7043, dtype: object

In [86]:
df['Cantidad_Servicios'] = df[servicios].sum(axis=1)
df.sample(3)

ID_Cliente  Evasion  Genero  Adulto_Mayor Tiene_Pareja  \
5311  7250-EQKIY        1  Female             0            1   
5377  7341-LXCAF        1    Male             0            1   
6858  9469-WEJBT        0    Male             0            1   

     Tiene_Dependientes  Antiguedad_Meses Servicio_Telefonico  \
5311                  1                 1                   1   
5377                  0                 4                   1   
6858                  0                37                   1   

     Multiples_Lineas Tipo_Internet  ... Streaming_TV Streaming_Peliculas  \
5311                0           DSL  ...            0                   0   
5377                1   Fiber optic  ...            0                   0   
6858                0   Fiber optic  ...            0                   1   

       Tipo_Contrato Factura_Digital       Metodo_Pago Cargos_Mensuales  \
5311  Month-to-month               0      Mailed check            50.45   
5377  Month-to-month               1  Electronic check            74.65   
6858  Month-to-month               1  Electronic check            90.35   

     Cargos_Totales Cargos_Diarios Segmento_Gasto_Diario  Cantidad_Servicios  
5311          50.45           1.68                   Low                   2  
5377         301.40           2.49                Medium                   2  
6858        3419.30           3.01                  High                   4  

[3 rows x 24 columns]

In [91]:
# Agrupamos por número de servicios y calculamos la tasa de evasión
tasa_evasion = df.groupby('Cantidad_Servicios')['Evasion'].mean().reset_index()
tasa_evasion

Cantidad_Servicios   Evasion
0                   0  0.437500
1                   1  0.211052
2                   2  0.328283
3                   3  0.364767
4                   4  0.313449
5                   5  0.255507
6                   6  0.224852
7                   7  0.124051
8                   8  0.052885

In [94]:
# Convertimos a porcentaje
tasa_evasion['Evasion_Porcentaje'] = (tasa_evasion['Evasion'] * 100).round(1)
tasa_evasion

Cantidad_Servicios   Evasion  Evasion_Porcentaje
0                   0  0.437500                43.8
1                   1  0.211052                21.1
2                   2  0.328283                32.8
3                   3  0.364767                36.5
4                   4  0.313449                31.3
5                   5  0.255507                25.6
6                   6  0.224852                22.5
7                   7  0.124051                12.4
8                   8  0.052885                 5.3

In [95]:
# Creamos el gráfico de barras
fig = px.bar(
    tasa_evasion,
    x='Cantidad_Servicios',
    y='Evasion_Porcentaje',
    text='Evasion_Porcentaje',
    title='Porcentaje de evasión según cantidad de servicios contratados',
    labels={
        'Cantidad_Servicios': 'Número de servicios contratados',
        'Evasion_Porcentaje': 'Evasión (%)'
    },
    color='Evasion_Porcentaje',
    color_continuous_scale='Reds'
)

fig.update_traces(
    textposition='outside'
)
fig.update_layout(
    width=800,
    height=500,
    yaxis=dict(range=[0, tasa_evasion['Evasion_Porcentaje'].max()*1.15]),
    plot_bgcolor='rgba(245,248,250,1)',
    font_family='Arial',
    font_size=14,
    title_font_size=18,
    showlegend=False
)

fig.show()


## 💡 Insights clave sobre evasión de clientes

Sin servicios → riesgo alto: 43.8% evadieron. Incentivar el primer servicio y comunicación personalizada para aumentar retención.

Primer servicio → gran impacto: Con 1 servicio, la evasión cae a 21.1%. Ofrecer experiencias iniciales atractivas y monitorear satisfacción.

2-3 servicios → pico de evasión: 32.8%-36.4%. Analizar combinaciones de servicios, mejorar comunicación de valor y recoger feedback específico.

4+ servicios → alta fidelización: La evasión baja a 5.3%. Incentivar cross-selling y programas de fidelidad para clientes premium.

## 🔑 Insight estratégico:

La primera venta es crítica, los servicios intermedios requieren atención y la fidelización aumenta con más servicios, haciendo retención y upselling muy rentables.